### PR TITLE
[Release/7.0] Updating inbox source generators to Roslyn 4.4 and removing polyfill approach

### PR DESCRIPTION
**Pull request by joperezr**
Fixes https://github.com/dotnet/runtime/issues/70754

### Description

This PR is upgrading all of the 7.0 inbox source generators (Json one is pending and I will add a commit that addresses that one too in a bit) to use Roslyn version 4.4 (version picked is the roslyn included in SDK 7.0.100-rc.1.22425.9), which allows us to remove the polyfill approach from them and instead use the inbox Roslyn APIs for finding an attribute in a performant manner. Additionally, this PR is also moving the repo to use that same roslyn version for the compilers used to build the repo.

### Customer Impact

After this change goes in, the only way to consume dotnet/runtime (without warnings or disabled analyzers) will be  by using a Roslyn4.4 compiler. These source generators are on by default when targeting dotnet 7.0, so all projects automatically pass them in to the compiler. Since these source generators will now depend on Roslyn 4.4, then the consuming app will need to use the Roslyn 4.4 compilers too, which can be done via using any of the following:

- Building using an SDK which is .NET 7 RC1 or older.
- Building using Visual Studio 17.4 preview 1 or older.
- Manually referencing Microsoft.NET.Compilers.Toolset package version  4.4.0-2.22423.18 or older

### Regression

No.

### Risk

Medium, since we are essentially changing the compilers that get used to build all of the repo, however, note that this needs to be done anyway since we want to ship .NET 7 with pretty much the final version of the compilers. This is also in a sense reducing risk, as it is removing from the inbox source generators the dependency to Roslyn API that was not fully tested and only source-copied. After this, we will now instead depend on the Roslyn tested bits.

### CODE REVIEW

**Comment by joperezr** [comment]
FYI: @jaredpar @marcpopMSFT

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs [hunk 1]
```diff
@@ -50,7 +49,7 @@ public override void Initialize(AnalysisContext context)
                 context =>
                 {
                     // Nothing to do if the LibraryImportAttribute is not in the compilation
-                    INamedTypeSymbol? libraryImportAttrType = context.Compilation.GetBestTypeByMetadataName(TypeNames.LibraryImportAttribute);
+                    INamedTypeSymbol? libraryImportAttrType = context.Compilation.GetTypeByMetadataName(TypeNames.LibraryImportAttribute);
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs:52]: i think you still want GetBestTypeByMetadataName.  That's a useful helper that solves important problems like when tehre may be multiple types in metadata with this name.

File: src/libraries/System.Private.CoreLib/gen/System.Private.CoreLib.Generators.csproj [hunk 2]
```diff
@@ -13,16 +13,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Compile Include="$(CommonPath)Roslyn\GetBestTypeByMetadataName.cs" Link="Common\Roslyn\GetBestTypeByMetadataName.cs" />
```
Comment (CyrusNajmabadi) [inline src/libraries/System.Private.CoreLib/gen/System.Private.CoreLib.Generators.csproj:16]: keep this guy?

**Comment by CyrusNajmabadi** [comment]
Done with pass.

File: Directory.Build.props [hunk 3]
```diff
@@ -299,7 +299,7 @@
     <Features>strict;nullablePublicOnly</Features>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <!-- Warnings to always disable -->
-    <NoWarn>$(NoWarn),CS8969</NoWarn>
+    <NoWarn>$(NoWarn);CS8969;IDE0031;IDE0074</NoWarn>
```
[the remark this answers is not in the document]
Comment (joperezr) [reply Directory.Build.props:302]: Oops, forgot to add comments for these. These are new compiler warnings that are logged for null checks that can be simplified, I could have addressed those on all offending files, but given this is a 7.0 PR, I wanted to minimize the changes to minimize the risk.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs (hunk 1 above)
Comment (joperezr) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs:52]: Is that API present in some other version of Roslyn? Or is that just not in? I thought the plan was to, if possible, not have to ship Roslyn bits as part of the generators that ship inbox.

**Comment by carlossanlop** [comment]
@danmoseley need an approval please.

**Comment by joperezr** [comment]
> need an approval please

Getting approval is fine now, but please do note that this is not yet ready to go in.

**Comment by carlossanlop** [comment]
> Getting approval is fine now, but please do note that this is not yet ready to go in.

Apologies. I missed the `NO MERGE` label.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs:52]: No, the `GetBestTypeByMetadataName` is not a public Roslyn API and is not planned to be. The definition of "Best" is nebulous, so Roslyn doesn't want to prescribe the policy there. However, the runtime teams and Roslyn have the same idea of what "best" means in this case, so we use this helper. We should keep this helper.

File: Directory.Build.props [hunk 4]
```diff
@@ -298,8 +298,12 @@
...
     <Features>strict;nullablePublicOnly</Features>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <!-- Warnings to always disable -->
-    <NoWarn>$(NoWarn),CS8969</NoWarn>
+    <!-- Warnings to always disable
+         CS8969 -> The private field 'class member' is never used 
+         IDE0031 -> Use null propagation 
+         IDE0074 -> Use compound assignment
+    -->
+    <NoWarn>$(NoWarn);CS8969;IDE0031;IDE0074</NoWarn>
```
Comment (eerhardt) [inline Directory.Build.props:306]: Why are we suppressing these here and not in the .editorconfig file?

**Comment by danmoseley** [comment]
This one is worth filling out the template for also. Eg., explaining how this is more than just code cleanup without visible effects.

> . For this reason, the plan is to, before merging this PR, merge a PR in dotnet/arcade that moves both the SDK and the Roslyn version up for all repos, and that way, after this goes in, upstack repos won't be broken.

I assume we want to do that anyway, so that we ultimately ship having built with the shipping SDK, more or less.

File: Directory.Build.props (hunk 4 above)
Comment (CyrusNajmabadi) [reply Directory.Build.props:306]: also, consider filing issues to reenable? up to you guys though.  but null-propagation and use-compound-assignment are both nice :)

**Comment by ViktorHofer** [comment]
One of the questions that Tactics might raise is why this isn't in main yet. Would it make sense to drive the required changes (including the SKD update) into main first?

**Comment by joperezr** [comment]
Main is already on a newer version of roslyn, that said it doesn't yet target it for the source generators specifically. I can certainly add a PR for that too.

**Comment by danmoseley** [comment]
Yes main first would be good - although, perhaps it would take some time before it flowed through and was testable in VS, so that would not increase coverage signficantly anytime soon (?)

**Comment by carlossanlop** [comment]
There's a merge conflict FYI.

**Comment by joperezr** [comment]
the new source generators that depend on roslyn 4.4 ship as part of the 7.0 targeting pack, so they will only be referenced if you are actually targeting .NET 7.0. Even if you use the new SDK, If you are targeting net6.0 then these generators should never be passed to the compiler.

**Comment by joperezr** [comment]
This is now ready for review.

@danmoseley I have filled the template now in the main description.

**Review by ViktorHofer — APPROVED** [review]

**Comment by joperezr** [comment]
Once PR https://github.com/dotnet/arcade/pull/10724 in arcade gets merged, all upstack repos will get an sdk bump and compiler bump automatically, which will allow them to ingest this change.

**Comment by carlossanlop** [comment]
> Once PR [dotnet/arcade#10724](https://github.com/dotnet/arcade/pull/10724) in arcade gets merged, all upstack repos will get an sdk bump and compiler bump automatically, which will allow them to ingest this change.

Ping me when ready to merge, please, @joperezr .

**Comment by danmoseley** [comment]
Approved -- we don't want to ship the way we are, this brings us into a regular state. Of course please watch the bits after they go in to ensure we spot any unexpected issue early.

**Comment by carlossanlop** [comment]
There's a merge conflict. Can one of you please resolve it, @CyrusNajmabadi or @joperezr?

**Comment by joperezr** [comment]
Yes, I'm working on this now. Turns out that the version that got merged into Arcade of the SDK had a different problem that I only discovered, will quickly push a PR to fix that, and then will go ahead and fix the changes here and merge

**Comment by joperezr** [comment]
@carlossanlop conflict is resolved and this should be ready to go. Feel free to merge as soon as CI passes 👍

**Comment by lewing** [comment]
```
  Starting:    JIT.HardwareIntrinsics.XUnitWrapper (parallel test collections = on, max threads = 2)
    JIT/HardwareIntrinsics/Arm/Sha256/Sha256_r/Sha256_r.sh [FAIL]
      Unhandled exception: System.IO.IOException: Connection timed out : 'Global\msbuild-server-launch-gkQydOqVnpHlKzivWKBFxHKmWmVPqCBlomIg710sl38'
         at System.Threading.Mutex.CreateMutexCore(Boolean initiallyOwned, String name, Boolean& createdNew)
         at Microsoft.Build.Experimental.MSBuildClient.TryLaunchServer()
         at Microsoft.Build.Experimental.MSBuildClient.Execute(CancellationToken cancellationToken)
         at Microsoft.Build.CommandLine.MSBuildClientApp.Execute(String[] commandLine, String msbuildLocation, CancellationToken cancellationToken)
         at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)
         at Microsoft.DotNet.Cli.Utils.MSBuildForwardingAppWithoutLogging.ExecuteInProc(String[] arguments)
         ```

**Comment by joperezr** [comment]
Yes, I'm working with @tannergooding to see if this is caused by the SDK update.

**Comment by carlossanlop** [comment]
@joperezr merging as requested.

**Pull request merged**